src/Plugins/Reminder.php

```diff
@@ -0,0 +1,357 @@
+                    'Set by ' . $user,
+                    'Seconds left: ' . $seconds
+                );
+            }
+
+            if(count($timeouts) !== count($reminders)){
+                return $this->chatClient->postMessage($command->getRoom(), $message);
+            }
+        });
+    }
+
+    /**
+     * Handle a command message
+     *
+     * According to http://www.strawpoll.me/11212318/r this plugin will support the following commands
+     * !!<reminder|remind> remind this <in|at> <time>
+     * !!at <time> remind this
+     * !!in <time> remind this
+     *
+     * @param Command $command
+     * @return Promise
+     */
+    public function handleCommand(Command $command): Promise
+    {
+        return resolve(function() use($command) {
+
+            if ($command->hasParameters() === false) {
+                return $this->chatClient->postMessage($command->getRoom(), self::USAGE);
+            }
+
+            /* $command->getParameter(0) can be: list | examples | flush | unset | <text> | <time> */
+            $textOrCommand = $command->getParameter(0);
+            $commandName = $command->getCommandName(); // <reminder|in|at>
+            switch ($commandName){
+                case 'reminder': case 'remind': break;
+                case 'in': case 'at': return yield $this->setReminder($command); break;
+            }
+
+            if ( count(array_diff($command->getParameters(), array($textOrCommand))) < 1 ){
+                switch ($textOrCommand){
+                    case 'list': return yield $this->getAllReminders($command); break;
+                    case 'examples': return yield $this->getExamples($command); break;
+                    case 'flush': return yield $this->flush($command);
+                }
+            }
+
+            if( $command->getParameter(0) === 'unset'
+                && $command->getParameter(1) !== null
+                && count($command->getParameters()) <= 2
+            ){ return yield $this->unset($command); }
+
+            return yield $this->setReminder($command);
+        });
+    }
+
+    public function apologize(ChatRoom $room, $timedOutReminders): \Generator
+    {
+        if(count($timedOutReminders) > 0){
+            foreach ($timedOutReminders as $key){
+                $value = yield $this->storage->get($key,$room);
+                $text = $value['text'];
+                $name = $value['username'];
+
+                if (null !== $pingableName = yield $this->chatClient->getPingableName($room, $name)) {
+                    $target = "@{$pingableName}";
+                    $reply = $target . " I guess I'm late: " . $text;
+                    yield $this->storage->unset($key, $room);
+                    return $this->chatClient->postMessage($room, $reply, PostFlags::ALLOW_PINGS);
+                }
+            }
+        }
+    }
+
+    public function reschedule(ChatRoom $room, array $upcomingReminders): \Generator
```

Comment:
Same as with `apologizeExpired `. Just pass in all reminders and let the method itself filter it.